src/main/java/org/prebid/server/bidder/criteo/CriteoBidder.java

```diff
@@ -0,0 +1,293 @@
+    }
+
+    private static String determineDeviceIdType(String deviceOs) {
+        if (deviceOs == null) {
+            return "unknown";
+        }
+        switch (deviceOs.toLowerCase()) {
+            case "ios":
+                return "idfa";
+            case "android":
+                return "gaid";
+            default:
+                return "unknown";
+        }
+    }
+
+    private static CriteoGdprConsent buildCriteoGdprConsent(BidRequest bidRequest, ExtRegs extRegs) {
+        final CriteoGdprConsent.CriteoGdprConsentBuilder criteoGdprConsentBuilder = CriteoGdprConsent.builder();
+        if (bidRequest.getUser() != null && bidRequest.getUser().getExt() != null) {
+            final ExtUser extUser = bidRequest.getUser().getExt();
+            criteoGdprConsentBuilder.consentData(extUser.getConsent());
+        }
+
+        if (extRegs != null && extRegs.getGdpr() != null) {
+            criteoGdprConsentBuilder.gdprApplies(extRegs.getGdpr() == 1);
+        }
+        return criteoGdprConsentBuilder.build();
+    }
+
+    private static MultiMap resolveHeaders(CriteoRequest criteoRequest) {
+        final MultiMap headers = HttpUtil.headers();
+        final CriteoUser criteoUser = criteoRequest.getUser();
+
+        if (criteoUser == null) {
+            return headers;
+        }
+        HttpUtil.addHeaderIfValueIsNotEmpty(
+                headers, HttpUtil.COOKIE_HEADER, String.format("uid=%s", criteoUser.getCookieId()));
+        HttpUtil.addHeaderIfValueIsNotEmpty(headers, HttpUtil.X_FORWARDED_FOR_HEADER, criteoUser.getIp());
+        HttpUtil.addHeaderIfValueIsNotEmpty(headers, HttpUtil.X_FORWARDED_FOR_HEADER, criteoUser.getIpV6());
+        HttpUtil.addHeaderIfValueIsNotEmpty(headers, HttpUtil.USER_AGENT_HEADER, criteoUser.getUserAgent());
+        return headers;
+    }
+
+    @Override
+    public Result<List<BidderBid>> makeBids(HttpCall<CriteoRequest> httpCall, BidRequest bidRequest) {
+        try {
+            final CriteoResponse criteoResponse
+                    = jsonMapper.decodeValue(httpCall.getResponse().getBody(), CriteoResponse.class);
+            return Result.withValues(extractBidsFromResponse(criteoResponse));
+        } catch (DecodeException e) {
+            return Result.withError(BidderError.badServerResponse(e.getMessage()));
+        }
+    }
+
+    private static List<BidderBid> extractBidsFromResponse(CriteoResponse criteoResponse) {
+        return criteoResponse.getSlots().stream()
+                .map(CriteoBidder::slotToBidderBid)
+                .collect(Collectors.toList());
+    }
+
+    private static BidderBid slotToBidderBid(CriteoResponseSlot slot) {
+        return BidderBid.of(slotToBid(slot), BidType.banner, slot.getCurrency());
+    }
+
+    private static Bid slotToBid(CriteoResponseSlot slot) {
+        return Bid.builder()
+                .id(slot.getId())
+                .impid(slot.getImpId())
+                .price(BigDecimal.valueOf(slot.getCpm()))
```

Comment:
You'll rise NPE if `slot.getCpm()` is null.

Reply:
Done